catgrad-core/src/interpreter/backend/candle.rs

```diff
@@ -0,0 +1,636 @@
+use super::super::types::*;
+use crate::category::core::{Dtype, Shape};
+use crate::interpreter::backend::{Backend, BackendError, NdArray};
+use candle_core::{DType, Device, Tensor};
+
+// ============================================================================
+// CANDLE BACKEND ARCHITECTURE EXPLANATION
+// ============================================================================
+//
+// This backend follows a two-layer architecture pattern common in Rust:
+//
+// 1. **`CandleTensor` - The Data Container**
+//    - Wrapper around `candle_core::Tensor`
+//    - Implements the `NdArray<D>` trait (required by the Backend trait)
+//    - Adds `PartialEq` implementation (missing from the underlying Tensor)
+//    - Provides type safety and API consistency with other backends
+//
+// 2. **`CandleBackend` - The Operations Provider**
+//    - Manages device state (CPU, GPU, Metal, etc.)
+//    - Implements the `Backend` trait with all operations (add, mul, matmul, etc.)
+//    - Handles device-specific configuration and state
+//
+// ARCHITECTURE PATTERN:
+//
+// Backend (Operations) ──→ Tensor (Data)
+//      ↓                      ↓
+// CandleBackend         CandleTensor
+//   - device              - Tensor
+//   - operations          - PartialEq
+//   - configuration       - NdArray trait
+//
+// COMPARISON WITH NDARRAY BACKEND:
+//
+// The ndarray backend is simpler because:
+// - `ArrayD<D>` already implements `PartialEq` and other required traits
+// - No device management needed (always CPU)
+// - No additional configuration required
+//
+// Candle needs both structs because:
+// - Device Management: Need to track which device tensors are on
+// - Missing Traits: `candle_core::Tensor` doesn't implement `PartialEq`
+// - Custom Behavior: Need to add our own methods and error handling
+//
+// This design follows the Single Responsibility Principle:
+// - `CandleTensor`: Manages tensor data and implements data-related traits
+// - `CandleBackend`: Manages operations and device state
+//
+// ============================================================================
+
+#[derive(Clone, Debug)]
+pub struct CandleTensor(pub Tensor);
+
+#[derive(Clone, Debug)]
+pub struct CandleBackend {
+    device: Device,
+}
+
+impl CandleBackend {
+    pub fn new() -> Self {
+        Self {
+            device: Device::Cpu,
+        }
+    }
+
+    pub fn with_device(device: Device) -> Self {
+        Self { device }
+    }
+}
+
+impl Default for CandleBackend {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Backend for CandleBackend {
+    type NdArray<D: HasDtype> = CandleTensor;
+
+    fn scalar<D: HasDtype>(&self, d: D) -> Self::NdArray<D> {
```

Comment:
IMO the `Backend` trait needs to be refactored so `scalar`, `zeros` and `ndarray_from_slice` can be implemented without `unsafe`. These are fine for now, but I've created #188 to track- we can change these later.